client/wfm/src/service/WfmService.ts

```diff
@@ -0,0 +1,241 @@
+        const stepResult = {
+          step,
+          submission,
+          status: STATUS.COMPLETE,
+          timestamp: new Date().getTime(),
+          submitter: profileData.id
+        };
+
+        // The result needs to be updated with the latest step results
+        workorderResult.stepResults = workorderResult.stepResults || {};
+        workorderResult.stepResults[step.code] = stepResult;
+        workorderResult.status = this.checkStatus(workorder, workflow, workorderResult);
+        workorderResult.nextStepIndex = this.executeStepReview(workflow.steps, workorderResult).nextStepIndex;
+
+        return this.resultService.update(workorderResult).then(function() {
+          return {
+            workorder,
+            workflow,
+            result: workorderResult,
+            nextStepIndex: workorderResult.nextStepIndex,
+            step: workorderResult.nextStepIndex > -1 ? workflow.steps[workorderResult.nextStepIndex] : workflow.steps[0]
+          };
+        });
+      });
+    });
+  }
+
+  protected createNewResult(workorderId, assignee) {
+    return this.resultService.create({
+      id: shortid.generate(),
+      status: STATUS.NEW_DISPLAY,
+      nextStepIndex: 0,
+      workorderId,
+      assignee,
+      stepResults: {}
+    });
+  }
+
+  /**
+   * Checks each of the result steps to determine if the workflow is complete,
+   * and if not, what is the next step in the workflow to display to the user.
+   */
+  protected executeStepReview(steps: Step[], result: WorkOrderResult) {
+    // tslint:disable-next-line:max-line-length
+    // See https://github.com/feedhenry-raincatcher/raincatcher-workflow/blob/b515e8acefad4bc50a7cc281863e2176c8babbed/lib/client/workflow-client/workflowClient.js
+    let nextStepIndex = 0;
+    let complete = false;
+
+    // If there is no result, then the first step is the next step.
+    if (result && !_.isEmpty(result.stepResults)) {
+      nextStepIndex = _.findIndex(steps, function(step) {
+        // The next incomplete step is the step with no entry or it's not complete yet.
+        return !result.stepResults || !result.stepResults[step.code] ||
+          result.stepResults[step.code].status !== STATUS.COMPLETE;
+      });
+
+      if (nextStepIndex === -1) {
+        complete = true;
+        nextStepIndex = steps.length;
+      }
+    }
+    return {
+      nextStepIndex,
+      complete // false means some steps are "pending"
+    };
+  }
+
+  /**
+   * Checking the status of a workorder
+   *
+   * @param workorder  - The workorder to check status
+   * @param workflow   - The workflow to check status
+   * @param result     - The result to check status
+   */
+  protected checkStatus(workorder: WorkOrder, workflow: WorkFlow, result: WorkOrderResult): STATUS {
+    let status;
+    const stepReview = this.executeStepReview(workflow.steps, result);
+    if (stepReview.nextStepIndex >= workflow.steps.length - 1 && stepReview.complete) {
```

Comment:
That doesn't sound like object oriented aproach. Maybe we can refactor this into separate class. StatusManager and allow people to override this if different statuses are done. I will want to do this also for `nextStepIndex` logic. It looks like nextStepIndex is used only for this specific logic, I believe we can also extract it.

To not overload you with this it can be done as separate PR, but I think we need that for our extensibility story.